libpromises/evalfunction.c

```diff
@@ -6659,6 +6659,189 @@ static FnCallResult FnCallReadData(ARG_UNUSED EvalContext *ctx,
+    }
+
+    char const *path = RlistScalarValue(args);
+    bool const has_heading = BooleanFromString(RlistScalarValue(args->next));
+    size_t const class_index = IntFromString(RlistScalarValue(args->next->next));
+    Rlist const *sort_arg = args->next->next->next;
+
+    FILE *csv_file = safe_fopen(path, "r");
+    if (csv_file == NULL)
+    {
+        Log(LOG_LEVEL_ERR,
+            "%s: Failed to read file %s: %s",
+            fp->name, path, GetErrorStrFromCode(errno));
+        return FnFailure();
+    }
+
+    Seq *heading = NULL;
+    JsonElement *json = JsonArrayCreate(50);
+    char *line;
+    size_t num_columns = 0;
+
+    // Current line number, for debugging
+    size_t line_number = 0;
+
+    while ((line = GetCsvLineNext(csv_file)) != NULL)
+    {
+        ++line_number;
+        if (line[0] == '#')
+        {
+            Log(LOG_LEVEL_DEBUG, "%s: Ignoring comment at line %zu",
+                fp->name, line_number);
+            free(line);
+            continue;
+        }
+
+        Seq *list = SeqParseCsvString(line);
+        free(line);
+        if (list == NULL)
+        {
+            Log(LOG_LEVEL_WARNING,
+                "%s: Failed to parse line %zu, line ignored.",
+                fp->name, line_number);
+            continue;
+        }
+
+        if (SeqLength(list) == 1 &&
+            strlen(SeqAt(list, 0)) == 0)
+        {
+            Log(LOG_LEVEL_DEBUG,
+                "%s: Found empty line at line %zu",
+                fp->name, line_number);
+            SeqDestroy(list);
+            continue;
+        }
+
+        if (num_columns == 0)
+        {
+            num_columns = SeqLength(list);
+
+            if (class_index >= num_columns)
+            {
+                Log(LOG_LEVEL_ERR,
+                    "%s: Class expression index is out of bounds. Row "
+                    " length %zu, index %zu",
+                    fp->name, num_columns, class_index);
+                FnFailure();
+            }
+        }
+        else if (num_columns != SeqLength(list))
+        {
+            Log(LOG_LEVEL_WARNING,
+                "%s: Line %zu has incorrect amount of elements, "
+                "%zu instead of %zu. Line ignored.",
+                fp->name, line_number, SeqLength(list), num_columns);
+            SeqDestroy(list);
+            continue;
+        }
+
+        // First parsed line is set to be heading if has_heading is true
+        if (has_heading && heading == NULL)
+        {
+            Log(LOG_LEVEL_DEBUG, "%s: Found header at line %zu",
+                fp->name, line_number);
+            heading = list;
+            SeqRemove(heading, class_index);
+        }
+        else
+        {
+            if (!IsDefinedClass(ctx, SeqAt(list, class_index)))
+            {
+                SeqDestroy(list);
+                continue;
+            }
+
+            SeqRemove(list, class_index);
+            JsonElement *json_object = JsonObjectCreate(num_columns);
```

Comment:
I think a better name than the generic `json_object` could make the code that follows more readable and self-explanatory.